src/terrasoApi/utils.ts

```diff
@@ -20,3 +20,24 @@ export const collapseConnectionEdges = <T>(connection: {
 }): T[] => {
   return connection.edges.map(({ node }) => node);
 };
+
+export function collapseFields<
+  Input extends {},
+  Output extends { [Property in keyof Partial<Output>]: Output[Property] },
+>(changes: {
+  [Property in keyof Partial<Output>]: (inp: Input) => Output[Property];
+}): (input: Input) => Output {
+  return (input: Input) =>
+    (Object.keys(input) as (keyof Output)[]).reduce((output, field) => {
+      if (field in changes) {
+        output[field] = changes[field](input);
+      } else {
+        if (field in input) {
+          output[field] = input[field as unknown as keyof Input] as any;
+        } else {
+          throw new Error(`field ${String(field)} not mapped to output`);
+        }
+      }
+      return output;
+    }, {} as Output);
+}
```

Comment:
i really like the idea of this function! also it is melting my brain! 

by inspection i'm not sure if it does the right thing? in particular, the if branch `if (field in input)` on line 35 will always be true, and because the fields are coming from the input any functions in `changes` that aren't in the input wouldn't get called because of the branch in line 32. am i reading this right? 

i also am somewhat flabbergasted that Output can appear in its own extend clause??

i'm not sure if typescript is able to really grok maps of types like this. i'm curious if a builder pattern would work better? no idea if this would work my brain is soup

```typescript
const collapseFields = <Field extends string, Input, Output>(field: Field, collapseFn: (_: Input) => Output) => {
  const apply = (input: Input) => Object.assign(_.omit(field, input), collapseFn(input));
  return {
    apply,
    then: <Field extends string, Output>(field: Field, collapseFn: (_: ReturnType<typeof apply>) => Output) => {
      collapseFields(field, inp => collapseFn(apply(inp)));
    },
    finalize: <T extends Output>() => apply,
  };
};

const collapseProjectFields = collapseFields('group', (proj: ProjectDataFragment) => ({userCount: proj.group.memberships.totalCount}))
  .then('updatedAt', ({updatedAt}) => ({updatedAt: new Date(updatedAt)}))
  .then('siteSet', ({siteSet: {totalCount}}) => ({siteCount: totalCount}))
  .finalize<Project>();
```